### PR TITLE
Adds decoration to the verifier errors.

### LINKED ISSUE #68: Prettier verifier errors
The verifier (#3) is beginning to find real bugs now. Its method for displaying verifier errors could be improved. Here's an example of the current output:

```
$ target/debug/cton-util test filetests/isa/riscv/expand-i32.cton 
FAIL filetests/isa/riscv/expand-i32.cton: inst1: arg 0 (v0) has type i16, must match function signature of i32
inst1: return v0

function large_imm(i16) -> i32 {
ebb0(vx0: i16):
    v0 = iadd_imm vx0, 0x3b9a_ca00
    return v0
}
```

The reference to `inst1` is cryptic because instruction references don't appear in the textual IL, so we print the instruction itself in the first line. This is not a great solution.

It would be much better to underline the offending instruction, like the filecheck tests do:

```
FAIL filetests/isa/riscv/expand-i32.cton: inst1: arg 0 (v0) has type i16, must match function signature of i32
function large_imm(i16) -> i32 {
ebb0(vx0: i16):
    v0 = iadd_imm vx0, 0x3b9a_ca00
    return v0
    ^~~~~~~~~
verifier: arg 0 (v0) has type i16, must match function signature of i32
}
```

# Implementation suggestion

One possible implementation is to move the guts of `write_function` into a more general

```rust
pub fn decorate_function<WL: FnMut(AnyEntity, usize, &str) -> fmt::Result>(w: WF, func: &Function, isa: Option<&TargetIsa>) -> fmt::Result;
```

The `WL` closure would be called to write out each line with the arguments: The entity being defined on the line, the indentation to insert before the line, and the contents of the line itself.

### PR DESCRIPTION
**Pull request by carolinecullen**
example:

function %bad(i32) fast {
ebb0(v0: i32):
    brnz.i32 v0, ebb1
    return
    ^~~~~~

verifier inst1: Internal return not allowed with return_at_end=1

ebb1:
    trapz.i32 v0, user6
    return
}

Fixes #68

### CODE REVIEW

**Comment by bjorn3** [comment]
Also fixes #373

File: lib/codegen/src/print_errors.rs [hunk 1]
```diff
@@ -1,27 +1,68 @@
...
+        &mut |w, func, isa, inst, indent| pretty_function_error(w, func, isa, inst, indent, err),
+        &mut w,
+        func,
+        isa,
+    ).unwrap();
+    w
+}
+
+/// Pretty-print a function verifier error.
+pub fn pretty_function_error(
```
Comment (sunfishcode) [inline lib/codegen/src/print_errors.rs]: Would this function be useful outside of this file? If not, it'd be good to remove the `pub`.

File: lib/codegen/src/write.rs [hunk 2]
```diff
@@ -9,9 +9,36 @@ use packed_option::ReservedValue;
 use std::fmt::{self, Write};
 use std::string::String;
 
+pub fn write_function_plain(
```
Comment (sunfishcode) [inline lib/codegen/src/write.rs]: Same with the `pub` here.

File: lib/codegen/src/write.rs [hunk 3]
```diff
@@ -9,9 +9,36 @@ use packed_option::ReservedValue;
...
 pub fn write_function(w: &mut Write, func: &Function, isa: Option<&TargetIsa>) -> fmt::Result {
+    decorate_function(
+        &mut |w, func, isa, inst, indent| write_function_plain(w, func, isa, inst, indent),
+        w,
+        func,
+        isa,
+    )
+}
+
+pub fn decorate_function<
```
Comment (sunfishcode) [inline lib/codegen/src/write.rs]: It'd be good to have a brief documentation comment here describing what `decorate_function` does.

File: lib/codegen/src/write.rs [hunk 4]
```diff
@@ -141,7 +168,15 @@ pub fn write_ebb_header(
     writeln!(w, "):")
 }
 
-pub fn write_ebb(w: &mut Write, func: &Function, isa: Option<&TargetIsa>, ebb: Ebb) -> fmt::Result {
+pub fn write_ebb<
```
Comment (sunfishcode) [inline lib/codegen/src/write.rs]: For consistency with `decorate_function`, I think we should rename `write_ebb` to `decorate_ebb`.

(In the future we could add a `write_ebb` which would do what `write_function` does, passing in `write_function_plain`, if we were to need such a thing.)

**Comment by carolinecullen** [comment]
Added the changes from your comments @sunfishcode !

**Comment by sunfishcode** [comment]
Looks good!

**Pull request merged**